nixos/modules/services/networking/wireguard-networkd.nix

```
@@ -0,0 +1,207 @@
+    matchConfig.Name = name;
+    address = interface.ips;
+  };
+
+  cfg = config.networking.wireguard;
+
+  refreshEnabledInterfaces = filterAttrs (
+    name: interface: interface.dynamicEndpointRefreshSeconds != 0
+  ) cfg.interfaces;
+
+  generateRefreshTimer =
+    name: interface:
+    nameValuePair "wireguard-dynamic-refresh-${name}" {
+      partOf = [ "wireguard-dynamic-refresh-${name}.service" ];
+      wantedBy = [ "timers.target" ];
+      description = "Wireguard dynamic endpoint refresh (${name}) timer";
+      timerConfig.OnBootSec = interface.dynamicEndpointRefreshSeconds;
+      timerConfig.OnUnitInactiveSec = interface.dynamicEndpointRefreshSeconds;
+    };
+
+  generateRefreshService =
+    name: interface:
+    nameValuePair "wireguard-dynamic-refresh-${name}" {
+      description = "Wireguard dynamic endpoint refresh (${name})";
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
+      path = with pkgs; [
+        iproute2
+        systemd
+      ];
+      # networkd doesn't provide a mechanism for refreshing endpoints.
+      # See: https://github.com/systemd/systemd/issues/9911
+      # This hack does the job but takes down the whole interface to do it.
+      script = ''
+        ip link delete ${name}
+        networkctl reload
+      '';
+    };
+
+in
+{
+  meta.maintainers = [ lib.maintainers.majiir ];
+
+  options.networking.wireguard = {
+    useNetworkd = mkOption {
+      default = config.networking.useNetworkd;
+      defaultText = literalExpression "config.networking.useNetworkd";
+      type = types.bool;
+      description = ''
+        Whether to use networkd as the network configuration backend for
+        Wireguard instead of the legacy script-based system.
+
+        ::: {.warning}
+        Some options have slightly different behavior with the networkd and
+        script-based backends. Check the documentation for each Wireguard
+        option you use before enabling this option.
+        :::
+      '';
+    };
+  };
+
+  config = mkIf (cfg.enable && cfg.useNetworkd) {
+
+    # TODO: Some of these options may be possible to support in networkd.
+    #
+    # privateKey and presharedKey are trivial to support, but we deliberately
+    # don't in order to discourage putting secrets in the /nix store.
+    #
+    # generatePrivateKeyFile can be supported if we can order a service before
+    # networkd configures interfaces. There is also a systemd feature request
+    # for key generation: https://github.com/systemd/systemd/issues/14282
+    #
+    # preSetup, postSetup, preShutdown and postShutdown may be possible, but
+    # networkd is not likely to support script hooks like this directly. See:
+    # https://github.com/systemd/systemd/issues/11629
+    #
+    # socketNamespace and interfaceNamespace can be implemented once networkd
+    # supports setting a netdev's namespace. See:
+    # https://github.com/systemd/systemd/issues/11629
```

I assume you meant

```suggestion
    # https://github.com/systemd/systemd/issues/11103
```

here?

---

Oops, thanks! Fixed in https://github.com/NixOS/nixpkgs/pull/364046